File: src/org/jgll/util/hashing/LevelMap.java
package org.jgll.util.hashing;

import java.io.Serializable;



public class LevelMap<K extends Level, V> implements Serializable {

	private static final long serialVersionUID = 1L;

	private LevelSet<MapEntry<K, V>> set;
	
	public LevelMap(ExternalHasher<K> decomposer) {	
		set = new LevelSet<>(new MapEntryDecomposer(decomposer));
	}
	
	public LevelMap(int initalCapacity, ExternalHasher<K> decomposer) {
		set = new LevelSet<>(initalCapacity, new MapEntryDecomposer(decomposer));
	}
		
	public V put(K key, V value) {
		MapEntry<K, V> entry = set.add(new MapEntry<K, V>(key, value));
		if(entry == null) {
			return null;
		}
		return entry.v;
	}
	
	public V get(K key) {
		MapEntry<K, V> entry = set.get(new MapEntry<K, V>(key, null));
		if(entry != null) {
			return entry.getValue();
		}
		
		return null;
	}

	public void clear() {
		set.clear();
	}
	
	public static class MapEntry<K extends Level, V> implements Level {
		
		private K k;
		private V v;
		
		public MapEntry(K k, V v) {
			this.k = k;
			this.v = v;
		}

		public K getKey() {
			return k;
		}

		public V getValue() {
			return v;
		}

		public V setValue(V value) {
			this.v = value;
			return v;
		}
		
		@Override
		public boolean equals(Object obj) {
			if(this == obj) {
				return true;
			}
			
			if(! (obj instanceof MapEntry)) {
				return false;
			}
			
			@SuppressWarnings("unchecked")
			MapEntry<K, V> other = (MapEntry<K, V>) obj;
			
			return k.equals(other.k);
		}

		@Override
		public int getLevel() {
			return k.getLevel();
		}
	}
	
	public class MapEntryDecomposer implements ExternalHasher<MapEntry<K, V>> {

		private ExternalHasher<K> externalHasher;

		public MapEntryDecomposer(ExternalHasher<K> externalHasher) {
			this.externalHasher = externalHasher;
		}
		
		@Override
		public int hash(MapEntry<K, V> e, HashFunction f) {
			return externalHasher.hash(e.k, f);
		}
	}
	
}

File: src-scala/org/jgll/util/hashing/LevelMap.scala
package org.jgll.util.hashing

import java.io.Serializable
import org.jgll.util.hashing.hashfunction.HashFunction
import LevelMap._
//remove if not needed
import scala.collection.JavaConversions._

object LevelMap {

  class MapEntry[K <: Level, V](var k: K, var v: V) extends Level {

    def getKey(): K = k

    def getValue(): V = v

    def setValue(value: V): V = {
      this.v = value
      v
    }

    override def equals(obj: Any): Boolean = {
      if (!(obj.isInstanceOf[MapEntry[K,V]])) {
        return false
      }
      val other = obj.asInstanceOf[MapEntry[K, V]]
      k == other.k
    }

    override def getLevel(): Int = k.getLevel
  }
}

@SerialVersionUID(1L)
class LevelMap[K <: Level, V >: Null](decomposer: ExternalHasher[K], initalCapacity: Int = CuckooHashSet.DEFAULT_INITIAL_CAPACITY)
                                     (implicit val kman: Manifest[K], implicit val vman: Manifest[V]) extends Serializable {

  private var set: LevelSet[MapEntry[K, V]] = new LevelSet(new MapEntryExternalHasher(decomposer))

  def put(key: K, value: V): V = {
    val entry = set.add(new MapEntry[K, V](key, value))
    if (entry == null) {
      return null
    }
    entry.v
  }

  def get(key: K): V = {
    val entry = set.get(new MapEntry[K, V](key, null))
    if (entry != null) {
      return entry.getValue
    }
    null
  }

  def clear() {
    set.clear()
  }

  @SerialVersionUID(1L)
  class MapEntryExternalHasher(private var externalHasher: ExternalHasher[K])
      extends ExternalHasher[MapEntry[K, V]] {

    override def hash(e: MapEntry[K, V], f: HashFunction): Int = externalHasher.hash(e.k, f)

    override def equals(e1: MapEntry[K, V], e2: MapEntry[K, V]): Boolean = externalHasher.equals(e1.k, e2.k)
  }
}
